### PR TITLE
Implement HTTP request adapter

### LINKED ISSUE #19: `kiota_http`: HTTP request adapter

### PR DESCRIPTION
**Pull request by ricardoboss**
Fixes #19.

I used https://github.com/microsoft/kiota-http-dotnet as inspiration :D

For now, I have skipped all the middleware because I am not sure how to achieve that with Dart's HTTP library. Also I want to add tests, but that shouldn't block this PR.

### CODE REVIEW

File: packages/kiota_http/lib/src/http_client_request_adapter.dart [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    final option = requestInfo.getRequestOption<ResponseHandlerOption>();
+
+    return option?.responseHandler;
+  }
+
+  Future<void> _throwIfFailedResponse(
+    http.StreamedResponse response,
+    ErrorMappings? errorMapping,
+  ) async {
+    if (response.statusCode >= 200 && response.statusCode < 300) {
```
Comment (andrueastman) [inline packages/kiota_http/lib/src/http_client_request_adapter.dart:76]: This should probably change to <400 based on recent developments. Checkout https://github.com/microsoft/kiota-http-dotnet/issues/219 and related items.

File: packages/kiota_http/lib/src/http_client_request_adapter.dart [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+      _baseUrl = value;
+    }
+  }
+
+  Future<http.BaseRequest> _getMessageFromInfo(
+    RequestInformation requestInfo,
+  ) async {
+    final request = http.Request(requestInfo.httpMethod!.name, requestInfo.uri);
+
+    // TODO(ricardoboss): implement the rest of the method
```
Comment (andrueastman) [inline packages/kiota_http/lib/src/http_client_request_adapter.dart:39]: This means at the moment requests with bodies/headers can't be sent. yes?

Comment (ricardoboss) [reply packages/kiota_http/lib/src/http_client_request_adapter.dart:39]: Whoops! Yeah, I forgot to implement that...

File: packages/kiota_http/lib/src/http_client_request_adapter.dart (hunk 1 above)
Comment (ricardoboss) [reply packages/kiota_http/lib/src/http_client_request_adapter.dart:92]: Should I only add an exception for `304`? Or special handling for `3xx` in general?

Comment (ricardoboss) [reply packages/kiota_http/lib/src/http_client_request_adapter.dart]: For now, I guess its best to just not throw for `3xx`.

**Comment by ricardoboss** [comment]
I am not sure what `RetryCAEResponseIfRequired` in the C# implementation does. Seemingly it retries the request with a value received from the `WWW-Authenticate` header. Maybe we should also implement that, but I'd need some guidance there.

**Comment by ricardoboss** [comment]
Looks like adding middleware can be done in a similar way as its done with the dotnet client: just wrap the client in delegating clients.

I'll work on this branch a bit more and mark this PR as draft for now.

**Review by andrueastman — APPROVED** [review]

**Pull request merged**